Clarified some of the fast scan behavior (#1486)
o3de/o3de.org: content/docs/user-guide/assets/asset-processor/faster-scanning.md

toc: true
---

**Faster Scanning Mode** speeds up **Asset Processor’s** startup scan by using timestamps to track source asset changes before checking file hashes. This saves time when launching Asset Processor in all circumstances. You can switch scan modes at any time without restarting the Asset Processor, including during an asset scan. Asset Processor saves your preference between sessions.

With Faster Scanning enabled, the process is:
1. The Asset Processor compares the timestamp of the source asset's last modification time against the last time it processed that file. If the timestamp is identical, it skips processing that source asset.
1. If the timestamps were different, it then creates a hash of the contents of the source asset and compares that to the hash of the source asset the last time it was processed. If they match, it skips processing the source asset.

>     * There are a few reasons why time stamps may not match but the hash of the content matches:

Author: Thanks for all this feedback! I've just pushed an update that should have it all applied.

    * There are a few reasons why time stamps may not match but the hash of the content matches:
        * Some source control configurations may result in the same source asset having a different timestamp.
        * Some archival formats, like zip, will truncate timestamps, so a source asset that is placed into a zip file and removed will have a truncated timestamp.
        * Sometimes content creators may re-save a file without making a change.
1. If the timestamp and hash are different, then this starts following the standard asset processing process, calling create jobs on the source asset for all relevant builders.

When Faster Scanning is disabled, the process is:
1. The Asset Processor calls Create Jobs on every single source asset it tracks, for each builder that processes those source assets.
1. If the fingerprint of the Create Jobs result for the source asset on that builder matches the last time it was run, then Process Job is skipped for that builder.

> With Faster Scanning enabled or disabled, if the fingerprint of the builder or the file have changed, the associated job will also run again. The builder fingerprint is modified by the builder author when they change the logic for the builder, and need all jobs using that builder to re-run. The file fingerprint checks if any of the source or job dependencies have changed, and will cause the job to re-run if those upstream files have been modified.

Reviewer: ```suggestion
With Faster Scanning enabled or disabled, if the fingerprint of the builder or the file has changed, the associated job will also run again. The builder fingerprint is modified by the builder author when they change the logic for the builder, and all jobs using that builder must be re-run. The file fingerprint checks if any source or job dependencies have changed and forces the job to re-run if those upstream files have been modified.
```

With Faster Scanning enabled, the process is:
1. The Asset Processor compares the timestamp of the source asset's last modification time against the last time it processed that file. If the timestamp is identical, it skips processing that source asset.
1. If the timestamps were different, it then creates a hash of the contents of the source asset and compares that to the hash of the source asset the last time it was processed. If they match, it skips processing the source asset.
    * There are a few reasons why time stamps may not match but the hash of the content matches:
        * Some source control configurations may result in the same source asset having a different timestamp.
        * Some archival formats, like zip, will truncate timestamps, so a source asset that is placed into a zip file and removed will have a truncated timestamp.
        * Sometimes content creators may re-save a file without making a change.

> 1. If the timestamp and hash are different, then this starts following the standard asset processing process, calling create jobs on the source asset for all relevant builders.

Reviewer 2: This is much better, the only thing I would add is that there's a final requirement that the builder's fingerprint and the file fingerprint haven't changed.  The file fingerprint is really just checking the dependencies haven't changed, so you can just say its checking the source/job dependencies haven't changed.

![The Faster Scanning Mode settings in Asset Processor](/images/user-guide/assets/pipeline/asset-processor-interface-fast-scan.png)

Faster Scanning Mode is on by default for Asset Processor GUI, but off by default for Asset Processor Batch.

To disable Faster Scanning Mode in the GUI, do the following:
1. Choose the Tools tab in Asset Processor.
1. Uncheck the **Faster Scanning Mode** option.

{{< note >}}

> When Faster Scanning Mode is disabled, file hashes are used to check for updated source assets and source dependencies. This might significantly increase analysis time.

Reviewer 2: I'd remove the "this might significantly increase analysis time" bit, "faster scanning mode" has always been shown to be faster (and the hash algorithm we use is super fast anyway)

Author: Fixed

{{< note >}}
If you are having issues with the Asset Cache, performing a full scan might resolve the issues. If a full scan does not repair the Asset Cache, you can rebuild the entire Asset Cache by deleting the `Cache` directory in your project. If you’re an engineer making BuilderSDK-based Asset Builders, deleting the cache is not recommended.
{{< /note >}}

> Deleting the Asset Cache to work around or solve a problem should be treated as a last resort solution. The problem might not be resolved after the cache is rebuilt, or if it does you might end up losing more time to deleting and rebuilding the cache instead of getting the root problem fixed. It's recommended that you document what issue you're facing before a rebuild, share that information with your team, and wait until a pipeline engineer has a chance to investigate before you proceed to deleting the cache.

Reviewer: ```suggestion
Deleting the Asset Cache to work around or solve a problem should be treated as a last resort solution. The issue might not be resolved after the cache is rebuilt. Even if it is, you might lose more time deleting and rebuilding the cache than it requires to fix the root problem. Instead, it is recommended that you document the issue and share that information with your team.  Wait until a pipeline engineer can investigate the problem before deleting the cache.
```
